### PR TITLE
WIP - Honor umask settings when writing cache files

### LINKED ISSUE #80581: The vulnerability(CVE-2019-11244) is not completely eliminated
**What happened**:
The PR(https://github.com/kubernetes/kubernetes/pull/77874) seems have fixed  [CVE-2019-11244](https://nvd.nist.gov/vuln/detail/CVE-2019-11244). 

But I doubt the correction didn't fix completely. 

According to the description of [CVE-2019-11244](https://nvd.nist.gov/vuln/detail/CVE-2019-11244):
> In Kubernetes v1.8.x-v1.14.x, schema info is cached by kubectl in the location specified by --cache-dir (defaulting to $HOME/.kube/http-cache), written with world-writeable permissions (rw-rw-rw-). If --cache-dir is specified and pointed at a different location accessible to other users/groups, the written files may be modified by other users/groups and disrupt the kubectl invocation.

The cache file created by kubectl should not be modified by other users/groups, or, to be more accurate, the file can only be modified by the user who runs the process.

But after the correction, the user in the same group still can write the cache file.
In the correction we have made the two changes:
* The patch permission changed from `0755` to `0750`.  (I think it's ok. Allow user in the same group enter the path but can not modify the path, neither create nor delete a file in this path.  Refuse all other users.)
* The cache file permission changed from `0755` to `0660`. (This still allows the user in the same group modify the cache file.)

So, I think the file permission should be `0640`. The user in the same group can only read the file.

**What you expected to happen**:
The cache file can only be modified by the user who runs the process, and the users in the same group can only have read permission.

**How to reproduce it (as minimally and precisely as possible)**:
1. Create a path by root and set permission with `0755`:
 `# mkdir -m 0750 myPath0750`

2. Create a file in it by root and set permission with `0660`:
`# touch myPath0750/myFile0660`
`# chmod 0660 myPath0750/myFile0660`

3. Create a new user in root group:
`# useradd -G root -d /home/horen -m horen`
`# passwd horen`

4. Switch to new users and try to modify the file `myPath0750/myFile0660`.
You will see the new user can modify the file.

**Anything else we need to know?**:
I don't know if there a scenario for users in the same group to modify the cache file. Please let me know if there is.

**Environment**:
- Kubernetes version (use `kubectl version`): N/A
- Cloud provider or hardware configuration: N/A
- OS (e.g: `cat /etc/os-release`): CentOS 7.0
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by liggitt**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Modifies writing of cache files to honor umask settings

**Which issue(s) this PR fixes**:
Fixes #80581

**Does this PR introduce a user-facing change?**:
```release-note
Cache files written to `~/.kube/{cache,http-cache}` now honor umask settings for determining whether files are group-readable/writeable
```

/sig cli
/sig api-machinery

### CODE REVIEW

File: staging/src/k8s.io/client-go/discovery/cached/disk/cached_discovery.go [hunk 1]
```diff
@@ -181,17 +182,23 @@ func (d *CachedDiscoveryClient) writeCachedFile(filename string, obj runtime.Obj
...
-	_, err = f.Write(bytes)
+
+	// remove world-wide permissions before writing data to the file
+	info, err := os.Stat(f.Name())
 	if err != nil {
 		return err
 	}
-
-	err = os.Chmod(f.Name(), 0660)
+	err = os.Chmod(f.Name(), info.Mode().Perm()&0660)
```
Comment (RainbowMango) [inline staging/src/k8s.io/client-go/discovery/cached/disk/cached_discovery.go]: os.Create() will create file with mode 0666, here `info.Mode().Perm()&0660` changes the mode to `0660`.

So, seems nothing changes. 

what're the benefits?

Comment (liggitt) [reply staging/src/k8s.io/client-go/discovery/cached/disk/cached_discovery.go]: os.Create honors umask, so will at most create a file with 0666. Most non root users default to umask of 0022, which means os.Create produces files of 0644 for them. We drop the world permissions and they end up with 0640.

Comment (RainbowMango) [reply staging/src/k8s.io/client-go/discovery/cached/disk/cached_discovery.go]: Ok, I understand. 

That means the final permission will depend on the user umask, we just drop the world permission as a supplement.

A novel solution I think, just a little complicated. :)

Should we keep group-writeable for files in this case? Can you take an example?

File: staging/src/k8s.io/client-go/discovery/cached/disk/cached_discovery_test.go [hunk 2]
```diff
@@ -102,7 +102,17 @@ func TestNewCachedDiscoveryClient_PathPerm(t *testing.T) {
...
 	assert.NoError(err)
-	os.RemoveAll(d)
+
+	// expect default read/write permissions with world-permissions removed
+	assert.NoError(os.MkdirAll(d, os.FileMode(0750)))
+	umask, err := os.Create(filepath.Join(d, "umask.tmp"))
+	assert.NoError(err)
+	umaskInfo, err := os.Stat(umask.Name())
+	assert.NoError(err)
+	expectedFilePermission := umaskInfo.Mode().Perm() & 0660
```
Comment (RainbowMango) [inline staging/src/k8s.io/client-go/discovery/cached/disk/cached_discovery_test.go]: I suggest creating const variables for 0660 and 0750. 
E.g.
```
const (
pathMode = 0xxx
fileMode = 0xxx
)
```

**Comment by RainbowMango** [comment]
ping @sig-cli for review.

**Comment by BenTheElder** [comment]
needs rebase. bot missed it..

**Comment by clayvan** [comment]
@liggitt Can you take a look at this again?

**Comment by liggitt** [comment]
I will not be able to get back to this until after kubecon

**Pull request closed without merging**